Allow Twitter notifications to include media

MikeChristianson opened the pull request:
**Allow Twitter notifications to include media.** 

The notifier uses the Twitter-recommended async chunked media/upload approach and tries to convey the correct mime type of the media. The implementation is based on https://github.com/geduldig/TwitterAPI/blob/master/examples/upload_video.py.

```yaml
notify:
  - name: twitter
    platform: twitter
    consumer_key: …
    consumer_secret: …
    access_token: …
    access_token_secret: …

automation twitter:
  alias: Send a test message
  initial_state: False
  hide_entity: False
  trigger:
    platform: time
    seconds: 00
  action:
    - service: notify.twitter
      data:
        message: "Test at {{ now().strftime('%I:%M:%S %p') }}."
        data:
          media: /tmp/my-awesome-video.mp4
    - service: homeassistant.turn_off
      entity_id: automation.send_a_test_message
```

MikeChristianson committed:
Allow notifications to include media, with Twitter as the first implementation. The Twitter notifier uses the Twitter-recommended async chunked media/upload approach and tries to convey the correct mime type of the media. Twitter implementation based on https://github.com/geduldig/TwitterAPI/blob/master/examples/upload_video.py.

homeassistant commented:
Hi @MikeChristianson, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#8282).

Once you do that we will be able to review and accept this pull request.

Thanks!

balloob commented on homeassistant/components/notify/__init__.py:
```diff
@@ -46,12 +49,13 @@
 NOTIFY_SERVICE_SCHEMA = vol.Schema({
     vol.Required(ATTR_MESSAGE): cv.template,
     vol.Optional(ATTR_TITLE): cv.template,
+    vol.Optional(ATTR_MEDIA): cv.isfile,
```
Other notify platforms are using `ATTR_DATA` for it. That way if a platform requires extra metadata, we don't need to add extra fields. So please add it to Twitter via `ATTR_DATA`.

MikeChristianson replied:
I attempted that approach, initially, but was not successful. I'll give it another try and look for examples.

balloob commented on tests/testing_config/automations.yaml:
```diff
@@ -0,0 +1 @@
+[]
```
Please remove this file. We fixed the issue in our tests that left this artifact.

balloob commented on homeassistant/components/notify/twitter.py:
```diff
@@ -51,17 +54,96 @@ def __init__(self, consumer_key, consumer_secret, access_token_key,
-        """Tweet a message."""
+        """Tweet a message, optionally with media"""
+        media = kwargs.get('media', None)
+        media_id = self.upload_media(media)
```
Media path has to be allowed by `hass.config.is_allowed_path`

MikeChristianson replied:
@balloob Can I do this in `[notify/__init__.py](https://github.com/MikeChristianson/home-assistant/blob/81a3be65398fe3732df04bf1b0b767ec29f77fb3/homeassistant/components/notify/__init__.py#L54)` so it applies to everything? Or must it be done, here and each other notifier?

balloob commented on homeassistant/components/notify/twitter.py:
```diff
@@ -51,17 +54,96 @@ def __init__(self, consumer_key, consumer_secret, access_token_key,
+            resp = self.upload_media_init(media_type, total_bytes)
+
+            media_id = None
+            if resp.status_code not in range(200, 299):
```
This is very inefficient. You are now generating up to 99 values. Prefer

```python
if 199 > resp.status_code < 300:
```

balloob commented on homeassistant/components/notify/twitter.py:
```diff
@@ -51,17 +54,96 @@ def __init__(self, consumer_key, consumer_secret, access_token_key,
+
+    def upload_media(self, media_path=None):
+        """Upload media."""
+        if media_path:
```
Shouldn't this method not be called if `media_path` is not set ? 

Also, prefer a guard clause:

```python
if not media_path:
    return

media_type, _ = mimetipes.guess_type(…)
```

balloob commented on homeassistant/components/notify/twitter.py:
```diff
@@ -51,17 +54,96 @@ def __init__(self, consumer_key, consumer_secret, access_token_key,
+
+            media_id = None
+            if resp.status_code not in range(200, 299):
+                self.log_error_resp(resp)
```
Since media_id is going to be None, why not just `return None` and you can remove the `else`.

MikeChristianson committed:
Changes based on balloob's review:
balloob: "Please remove this file. We fixed the issue in our tests that left this artifact."
balloob: "…prefer a guard clause"
balloob: "This is very inefficient. You are now generating up to 99 values."
balloob: "Since media_id is going to be None, why not just return None and you can remove the else."

MikeChristianson committed:
balloob: "Other notify platforms are using ATTR_DATA for it. That way if a platform requires extra metadata, we don't need to add extra fields. So please add it to Twitter via ATTR_DATA."

MikeChristianson commented:
@balloob Please have another look, thanks.

balloob commented:
Awesome! 🐬  🎉

MikeChristianson commented:
Thanks, @balloob! 🎊 Glad I could help the community, too.

onerealfunnyguy commented:
Can you please update the component page to include the correct formatting for images, can send txt tweet no problems, yet with the above example using both a local path and http:// I can not get the notify to work once I add the include media code.
